### PR TITLE
Opencv updates

### PR DESCRIPTION
**Pull request by bvanessen**
Updated the number of variants available and grouped the variants according to OpenCV features.

### CODE REVIEW

File: var/spack/repos/builtin/packages/lbann/package.py [hunk 1]
```diff
@@ -51,8 +51,13 @@ class Lbann(CMakePackage):
...
     depends_on('mpi')
-    depends_on('hwloc')
-    depends_on('opencv@3.2.0: +openmp +core +highgui +imgproc +jpeg +png +tiff +zlib ~eigen', when='+opencv')
+    depends_on('hwloc ~pci ~libxml2')
+    # LBANN wraps OpenCV calls in OpenMP parallel loops, build without OpenMP
+    depends_on('opencv@3.2.0: +core +highgui +imgproc +jpeg +png +tiff +zlib '
+               '+fast-math ~lapack ~calib3d ~dnn ~features2d ~flann ~java ~ml '
+               '~python ~stitching ~superres ~ts ~video ~videostab ~videoio '
+               '~eigen ~pthreads_pf ~opencl ~opencl_svm ~openclamdfft '
+               '~openclamdblas', when='+opencv')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lbann/package.py]: Just to confirm, every one of these variants is absolutely required to build `lbann`? A lot of people use variants that they "prefer", but aren't actually required. We're trying to cut down on that as it restricts what combinations people can build.

**Comment by bvanessen** [comment]
Most of these variant flags are disabling parts of OpenCV that LBANN does not take advantage of this.  The reason to do this is to make OpenCV build faster. 

Brian C. Van Essen
vanessen1@llnl.gov
(w) 925-422-9300
(c) 925-290-5470

**Comment by adamjstewart** [comment]
> Most of these variant flags are disabling parts of OpenCV that LBANN does not take advantage of this.  The reason to do this is to make OpenCV build faster.

I would request that you not do this. Package dependencies are _requirements_, not _suggestions_. If someone already has `opencv` installed but with more features (like the defaults), then Spack would have to reinstall `opencv` to satisfy these overly tight constraints. If you want to disable those variants locally, you can change the defaults in your `packages.yaml`.

**Comment by bvanessen** [comment]
Following up on the previous comment about the number of variants in OpenCV that are on by default.  I agree in principle that it would be great to create a minimally restrictive set of variants that are required for any given package.  However, that requires packages to not turn on everything by default.  I added a bunch of variants to the OpenCV package and I set the sense of them to match what was originally in the OpenCV build environment, however, many optional things are turned on by default.  Additionally, my team has found a bug in building with OpenCV where the video options include a "silent" dependency on OpenMP and when we are building with clang OpenCV is not getting the correct OpenMP library.  Therefore it is necessary to turn off the video variants.

**Comment by bvanessen** [comment]
As a potential solution to some of this problem, I can change the defaults for the OpenCV variants to disable most of them by default. Please note that this will change the default spack build experience from the straight cmake environment.

**Comment by bvanessen** [comment]
I have updated the package variants to be off by default based on Todd's recommendation.  I looked at the dependencies and most of the packages that use OpenCV have explicitly positive dependencies and should work fine.  The CNTK package has no variants turned on and but I am not sure what it would need.  On the CNTK website it is asserted that OpenCV is only necessary for TensorBoard and image reader.

**Comment by scheibelp** [comment]
closing/reopening to restart tests

**Comment by scheibelp** [comment]
I saw it hinted at but I'll reiterate: with this PR, if you say `lbann` depends on `opencv ~opencl` (as an example - I'm aware this isn't in `lbann`'s package.py) it means that it is incompatible with `opencv +opencl`.

Breaking this down into components and turning them off by default helps `lbann` here but then it becomes complex for each other dependent of `opencv` to decide which variants to enable. If the sets of enabled variants for dependents do not match, then each dependent will build a separate opencv (to be clear this problem comes up when they are off by default). 

If you turned off a variant by default for optimization reasons related specifically to `lbann` then IMO it's worth making it true by default and updating the preferred variants in packages.yaml on your system. In this case, when someone doesn't want to think about which `opencv` components are required they build everything and all dependents use the same thing (in which case the build process takes longer but that's ok for some folks).

Examples of variants which likely should be off by default:

* Based on https://github.com/spack/spack/pull/6502#issuecomment-371992711 IMO it would make sense for the `video*` variants to default to off. Also for example it would make sense to let `openmp` remain off by default.
* Variants related to specific architectures (`powerpc`, `vsx`) should also be off by default (so they are fine as they are currently set in the PR)

Side note: in addition to the `ffmpeg` dependency should there also be an `mpi` dependency for `opencv+videoio`?

Does that seem agreeable?

File: var/spack/repos/builtin/packages/opencv/package.py [hunk 2]
```diff
@@ -93,44 +121,97 @@ class Opencv(CMakePackage):
...
-                'ON' if '+vtk' in spec else 'OFF')),
-            '-DBUILD_opencv_java:BOOL={0}'.format((
-                'ON' if '+java' in spec else 'OFF')),
+            '-DWITH_LAPACK={0}'.format((
+                'ON' if '+lapack' in spec else 'OFF')),
+            '-DENABLE_POWERPC={0}'.format((
+                'ON' if '+powerpc' in spec else 'OFF')),
+            '-DENABLE_VSX={0}'.format((
+                'ON' if '+vsx' in spec else 'OFF')),
+            '-DENABLE_FAST_MATH={}'.format((
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/opencv/package.py]: @bvanessen the `{}`-style formatting is not compatible with python 2.6, update it to `{0}`

**Comment by bvanessen** [comment]
So I don't know what to do with this package.  The make environment builds almost everything by default and including all modules and 3rd party components.  When I talked with @tgamblin he suggested that he would like to see packages build with the minimal set of variants in general and he was okay not following the existing set of default build options.  I can really go any way, but it is annoying to force everyone who wants to use OpenCV for simple things like image processing to either build everything, or cart around personal package.yaml files that have to be shared among team members to have an efficient build environment.

**Comment by tgamblin** [comment]
Hey all, for LBANN I think for now we should allow Brian to disable the variants he doesn't need to get a faster build.  We've done this in one other case for conduit -- it's typically for libraries that are in development and that aren't widely used.  It doesn't have *that* many reuse ramifications at the moment, because Spack doesn't aggressively reuse things at the moment anyway (we're working on that).

I have designs to be able to support (by September) something like `spack install --minimal` so that we can have people specify their variants the _right_ way (i.e. only specify what is _needed_), but also give users the ability to do an install with the fewest variants possible.  This would be a concretizer policy.

So for now, LBANN is small enough I think we should let Brian do what he wants _in lbann_, but maintain sensible defaults in opencv.  Let's revisit in September when `--minimal` is possible.  Thoughts?

**Comment by bvanessen** [comment]
I updated the OpenCV package so that it has the same sense on the variants as the previous version.  It should be good to merge.

**Comment by bvanessen** [comment]
@scheibelp Here is the latest update that we discussed.

**Comment by scheibelp** [comment]
Thanks!

**Comment by bvanessen** [comment]
Thank you/

Brian C. Van Essen
vanessen1@llnl.gov
(w) 925-422-9300
(c) 925-290-5470

**Pull request merged**